public/skin/frontend/base/default/js/opcheckout.js

```diff
@@ -23,211 +24,256 @@ Checkout.prototype = {
+
+        if (useForShippingYes && useForShippingYes.checked) {
             shipping.syncWithBilling();
-            $('opc-shipping').addClassName('allow');
+            opcShipping.classList.add('allow');
             this.gotoSection('shipping_method', true);
-        } else if (($('billing:use_for_shipping_no')) && ($('billing:use_for_shipping_no').checked)) {
-            $('shipping:same_as_billing').checked = false;
+        }
+        else if (useForShippingNo && useForShippingNo.checked) {
+            sameAsBilling.checked = false;
             this.gotoSection('shipping', true);
-        } else {
-            $('shipping:same_as_billing').checked = true;
+        }
+        else {
+            sameAsBilling.checked = true;
             this.gotoSection('shipping', true);
         }
-    },
+    }
 
-    setShipping: function() {
+    setShipping() {
         this.gotoSection('shipping_method', true);
-    },
+    }
 
-    setShippingMethod: function() {
+    setShippingMethod() {
         this.gotoSection('payment', true);
-    },
+    }
 
-    setPayment: function() {
+    setPayment() {
         this.gotoSection('review', true);
-    },
+    }
 
-    setReview: function() {
+    setReview() {
         this.reloadProgressBlock();
-    },
+    }
 
-    back: function(){
+    back() {
         if (this.loadWaiting) return;
-        //Navigate back to the previous available step
-        var stepIndex = this.steps.indexOf(this.currentStep);
-        var section = this.steps[--stepIndex];
-        var sectionElement = $('opc-' + section);
 
-        //Traverse back to find the available section. Ex Virtual product does not have shipping section
+        // Navigate back to the previous available step
+        let stepIndex = this.steps.indexOf(this.currentStep);
+        let section = this.steps[--stepIndex];
+        let sectionElement = document.getElementById(`opc-${section}`);
+
+        // Traverse back to find the available section. Ex Virtual product does not have shipping section
         while (sectionElement === null && stepIndex > 0) {
             --stepIndex;
             section = this.steps[stepIndex];
-            sectionElement = $('opc-' + section);
+            sectionElement = document.getElementById(`opc-${section}`);
         }
-        this.changeSection('opc-' + section);
-    },
 
-    setStepResponse: function(response){
+        this.changeSection(`opc-${section}`);
+    }
+
+    setStepResponse(response) {
         if (response.update_section) {
-            $('checkout-'+response.update_section.name+'-load').update(response.update_section.html);
+            const element = document.getElementById(`checkout-${response.update_section.name}-load`);
+            const range = document.createRange();
+            const fragment = range.createContextualFragment(response.update_section.html);
+            element.innerHTML = ''; // Clear existing content
+            element.appendChild(fragment);
```

Comment:
This is so `<script>` blocks execute, right? I created a helper function while removing ExtJS to do the same thing, but I actually like your version better.

Mine: https://github.com/justinbeaty/maho/blob/df443e0aff8aa46d1119de78aa6d4efecff3cf3a/public/js/varien/js.js#L71-L95

Yours with some slight modification:
```js
function updateElementHTML(targetEl, html, executeExternalScripts = false) {
    const range = document.createRange();
    const fragment = range.createContextualFragment(html);
    if (!executeExternalScripts) {
        fragment.querySelectorAll('script[src]').forEach(script => script.remove());
    }
    targetEl.replaceChildren(fragment);
}
```

The `executeExternalScripts` thing is because prototype by default doesn't execute `<script src="..."></script>` tags, so I wanted to mimic that behavior. Not sure any returned HTML from Magento contains external scripts though...

Reply:
should I put this new version in `js.js` as we did for the `stripTags`?

Reply:
Yeah that's where I put it while working on ExtJS since `js.js` is loaded on both frontend and backend. I have a feeling we will need to use it more once we eventually remove prototype from admin.

Reply:
> The `executeExternalScripts` thing is because prototype by default doesn't execute `<script src="..."></script>` tags, so I wanted to mimic that behavior. Not sure any returned HTML from Magento contains external scripts though...

in the checkout there are a lot of js returned via ajax (cause they're inside phtml files) so, if that's ok with you, I think that the expected behavior is to execute the js, would you prefer to have it disabled?

Reply:
@justinbeaty done it in https://github.com/MahoCommerce/maho/pull/71/commits/069c5785d2804866fb0886ea9f015d9f8dc6feaf

and probably, for security reasons, it's better to keep the executeScripts as disabled

Reply:
What I meant is that prototype's `Ajax.Updater` only executes inline scripts, i.e. `<script>doFoo()</script>` and not external scripts like `<script src="https://..."></script>`.

I just figured instead of outright denying external scripts, we would give a parameter to control it. But to keep compatibility with prototype, you actually want to remove that `true` parameter.

Maybe the function should be named better, like `updateElementHTMLandExecuteScripts`, because if you *don't* want to execute scripts you can just keep using `element.innerHTML = '...'`

Hope that makes sense.

Reply:
Ref from http://api.prototypejs.org/ajax/Ajax/Updater/index.html

![image](https://github.com/user-attachments/assets/8da0e574-6414-4505-a87f-51e5e52751c9)

Reply:
@justinbeaty makes perfect sense, done in https://github.com/MahoCommerce/maho/pull/71/commits/6a76ca55f2b547a83fb4d5ac165fc8e50ebed7fc

I've used the updateElementHtmlAndExecuteScripts following this guideline:
<img width="701" alt="Screenshot 2024-12-16 alle 17 48 29" src="https://github.com/user-attachments/assets/3ffa1116-eb86-41c3-94e8-d479d49b219a" />

Reply:
Yup, that function name is better 😃